File: priv/rust/tui/src/app/keys.rs
// Phase 3: Centralized keymap — wire when splitting update.rs key handlers
#![allow(dead_code)]

use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};

/// Key binding definition
pub struct KeyBinding {
    pub code: KeyCode,
    pub modifiers: KeyModifiers,
    pub help: &'static str,
}

impl KeyBinding {
    pub const fn new(code: KeyCode, modifiers: KeyModifiers, help: &'static str) -> Self {
        Self {
            code,
            modifiers,
            help,
        }
    }

    pub fn matches(&self, event: &KeyEvent) -> bool {
        event.code == self.code && event.modifiers.contains(self.modifiers)
    }
}

/// All key bindings — 22 bindings matching Go keymap
pub struct KeyMap {
    pub submit: KeyBinding,
    pub cancel: KeyBinding,
    pub quit_eof: KeyBinding,
    pub escape: KeyBinding,
    pub help: KeyBinding,
    pub page_up: KeyBinding,
    pub page_down: KeyBinding,
    pub scroll_top: KeyBinding,
    pub scroll_bottom: KeyBinding,
    pub scroll_up: KeyBinding,
    pub scroll_down: KeyBinding,
    pub half_page_up: KeyBinding,
    pub half_page_down: KeyBinding,
    pub toggle_expand: KeyBinding,
    pub toggle_thinking: KeyBinding,
    pub toggle_background: KeyBinding,
    pub toggle_sidebar: KeyBinding,
    pub tab: KeyBinding,
    pub clear_input: KeyBinding,
    pub new_session: KeyBinding,
    pub palette: KeyBinding,
    pub copy_message: KeyBinding,
}

impl Default for KeyMap {
    fn default() -> Self {
        Self {
            submit: KeyBinding::new(KeyCode::Enter, KeyModifiers::NONE, "submit"),
            cancel: KeyBinding::new(KeyCode::Char('c'), KeyModifiers::CONTROL, "cancel/quit"),
            quit_eof: KeyBinding::new(KeyCode::Char('d'), KeyModifiers::CONTROL, "quit"),
            escape: KeyBinding::new(KeyCode::Esc, KeyModifiers::NONE, "cancel"),
            help: KeyBinding::new(KeyCode::F(1), KeyModifiers::NONE, "help"),
            page_up: KeyBinding::new(KeyCode::PageUp, KeyModifiers::NONE, "page up"),
            page_down: KeyBinding::new(KeyCode::PageDown, KeyModifiers::NONE, "page down"),
            scroll_top: KeyBinding::new(KeyCode::Home, KeyModifiers::NONE, "scroll top"),
            scroll_bottom: KeyBinding::new(KeyCode::End, KeyModifiers::NONE, "scroll bottom"),
            scroll_up: KeyBinding::new(KeyCode::Char('k'), KeyModifiers::NONE, "scroll up"),
            scroll_down: KeyBinding::new(KeyCode::Char('j'), KeyModifiers::NONE, "scroll down"),
            half_page_up: KeyBinding::new(KeyCode::Char('u'), KeyModifiers::NONE, "half page up"),
            half_page_down: KeyBinding::new(
                KeyCode::Char('d'),
                KeyModifiers::NONE,
                "half page down",
            ),
            toggle_expand: KeyBinding::new(
                KeyCode::Char('o'),
                KeyModifiers::CONTROL,
                "expand/collapse",
            ),
            toggle_thinking: KeyBinding::new(
                KeyCode::Char('t'),
                KeyModifiers::CONTROL,
                "toggle thinking",
            ),
            toggle_background: KeyBinding::new(
                KeyCode::Char('b'),
                KeyModifiers::CONTROL,
                "background task",
            ),
            toggle_sidebar: KeyBinding::new(
                KeyCode::Char('l'),
                KeyModifiers::CONTROL,
                "toggle sidebar",
            ),
            tab: KeyBinding::new(KeyCode::Tab, KeyModifiers::NONE, "autocomplete"),
            clear_input: KeyBinding::new(
                KeyCode::Char('u'),
                KeyModifiers::CONTROL,
                "clear input",
            ),
            new_session: KeyBinding::new(
                KeyCode::Char('n'),
                KeyModifiers::CONTROL,
                "new session",
            ),
            palette: KeyBinding::new(
                KeyCode::Char('k'),
                KeyModifiers::CONTROL,
                "command palette",
            ),
            copy_message: KeyBinding::new(
                KeyCode::Char('y'),
                KeyModifiers::NONE,
                "copy last message",
            ),
        }
    }
}

File: priv/go/tui-v2/app/keys.go
package app

import "charm.land/bubbles/v2/key"

// KeyMap defines all global keybindings.
type KeyMap struct {
	// Global
	Submit  key.Binding
	Cancel  key.Binding
	QuitEOF key.Binding
	Escape  key.Binding
	Help    key.Binding

	// Navigation
	PageUp       key.Binding
	PageDown     key.Binding
	ScrollTop    key.Binding
	ScrollBottom key.Binding
	ScrollUp     key.Binding // k
	ScrollDown   key.Binding // j
	HalfPageUp   key.Binding // u
	HalfPageDown key.Binding // d

	// Toggles
	ToggleExpand     key.Binding
	ToggleThinking   key.Binding
	ToggleBackground key.Binding
	ToggleSidebar    key.Binding

	// Editor
	Tab        key.Binding
	ClearInput key.Binding

	// Commands
	NewSession key.Binding
	Palette    key.Binding

	// Copy
	CopyMessage key.Binding

	// Search
	Search key.Binding
}

// DefaultKeyMap returns the default keybindings.
func DefaultKeyMap() KeyMap {
	return KeyMap{
		Submit: key.NewBinding(
			key.WithKeys("enter"),
			key.WithHelp("enter", "submit"),
		),
		Cancel: key.NewBinding(
			key.WithKeys("ctrl+c"),
			key.WithHelp("ctrl+c", "cancel/quit"),
		),
		QuitEOF: key.NewBinding(
			key.WithKeys("ctrl+d"),
			key.WithHelp("ctrl+d", "quit"),
		),
		Escape: key.NewBinding(
			key.WithKeys("esc"),
			key.WithHelp("esc", "cancel"),
		),
		Help: key.NewBinding(
			key.WithKeys("f1"),
			key.WithHelp("F1", "help"),
		),
		PageUp: key.NewBinding(
			key.WithKeys("pgup"),
			key.WithHelp("pgup", "page up"),
		),
		PageDown: key.NewBinding(
			key.WithKeys("pgdown"),
			key.WithHelp("pgdn", "page down"),
		),
		ScrollTop: key.NewBinding(
			key.WithKeys("home"),
			key.WithHelp("home", "scroll top"),
		),
		ScrollBottom: key.NewBinding(
			key.WithKeys("end"),
			key.WithHelp("end", "scroll bottom"),
		),
		ScrollUp: key.NewBinding(
			key.WithKeys("k"),
			key.WithHelp("k", "scroll up"),
		),
		ScrollDown: key.NewBinding(
			key.WithKeys("j"),
			key.WithHelp("j", "scroll down"),
		),
		HalfPageUp: key.NewBinding(
			key.WithKeys("u"),
			key.WithHelp("u", "half page up"),
		),
		HalfPageDown: key.NewBinding(
			key.WithKeys("d"),
			key.WithHelp("d", "half page down"),
		),
		ToggleExpand: key.NewBinding(
			key.WithKeys("ctrl+o"),
			key.WithHelp("ctrl+o", "expand/collapse"),
		),
		ToggleThinking: key.NewBinding(
			key.WithKeys("ctrl+t"),
			key.WithHelp("ctrl+t", "toggle thinking"),
		),
		ToggleBackground: key.NewBinding(
			key.WithKeys("ctrl+b"),
			key.WithHelp("ctrl+b", "background"),
		),
		ToggleSidebar: key.NewBinding(
			key.WithKeys("ctrl+l"),
			key.WithHelp("ctrl+l", "toggle sidebar"),
		),
		Tab: key.NewBinding(
			key.WithKeys("tab"),
			key.WithHelp("tab", "autocomplete"),
		),
		ClearInput: key.NewBinding(
			key.WithKeys("ctrl+u"),
			key.WithHelp("ctrl+u", "clear input"),
		),
		NewSession: key.NewBinding(
			key.WithKeys("ctrl+n"),
			key.WithHelp("ctrl+n", "new session"),
		),
		Palette: key.NewBinding(
			key.WithKeys("ctrl+k"),
			key.WithHelp("ctrl+k", "command palette"),
		),
		CopyMessage: key.NewBinding(
			key.WithKeys("y", "c"),
			key.WithHelp("y/c", "copy message"),
		),
		Search: key.NewBinding(
			key.WithKeys("ctrl+f"),
			key.WithHelp("ctrl+f", "search chat"),
		),
	}
}
